### PR TITLE
Missing Tokens Script Update

### PR DESCRIPTION
**Pull request by bh2smith**
Now Using the official Dune API!

This script was used to generate [this PR](https://github.com/duneanalytics/spellbook/pull/1720)

Console logs look like this:

```
Fetching V1 missing tokens from 1317323
NFO dune_client.client waiting for query execution 01GES9BVY7MBMNDZF90CP22N4H to complete: ExecutionState.EXECUTING
Fetching V2 missing tokens from 1367051
INFO dune_client.client waiting for query execution 01GES90D7BQACRXQNXFQFR84B4 to complete: ExecutionState.PENDING (queue position: 7)
INFO dune_client.client waiting for query execution 01GES90D7BQACRXQNXFQFR84B4 to complete: ExecutionState.PENDING (queue position: 2)

Found 18 missing tokens on V1 and 17 on V2, fetching metadata...

V1 results:

\\x12b6893cE26Ea6341919FE289212ef77e51688c8	TAMA	18
\\x17e6BA2519B4d15199B6529dB340910Ae031b1B0	oneGIV	18
\\x220c1345fb076a6eF4daD9a9688e2b9b724528FF	UCHI	9
\\x24E79e946dEa5482212c38aaB2D0782F04cdB0E0	palStkAAVE	18
\\x3231Cb76718CDeF2155FC47b5286d82e6eDA273f	EURe	18
\\x3A9780555DAC7aD1c1B252f466ddaE09E5737F50	⸙͎۪۫	18
\\x467719aD09025FcC6cF6F8311755809d45a5E5f3	AXL	6
\\x4fdbcC7331a03B1E323Ccd8968966282950ad769	QMT	18
\\x6B0956258fF7bd7645aa35369B55B61b8e6d6140	LUCKY	8
\\x9625cE7753ace1fa1865A47aAe2c5C2Ce4418569	KAP	18
\\xb29eDfF8345b9d8039d8B9471C027D6e458a3C02	$OOO	18
\\xB43B2A25330ddA3366417a6a5F3b961154b6F99c	SUN	18
\\xc502002aEB1b9309FcCb016adF50507987Fc6C2b	GNFT	18
\\xC5bcc8BA3F33Ab0d64f3473e861BDC0685b19Ef5	$MECHA	18
\\xE1E44E6918c283a0481037a03056B416f4347B28	DAZU	9
\\xE5f55a3b74874531a99359b833b92866A6609f6B	ATS	4
\\xe9f84d418B008888A992Ff8c6D22389C2C3504e0	BASE	8
\\xfa372fF1547fa1a283B5112a4685F1358CE5574d	BDAMM	18

V2 results:

('0x12b6893ce26ea6341919fe289212ef77e51688c8', 'TAMA', 18),
('0x17e6ba2519b4d15199b6529db340910ae031b1b0', 'oneGIV', 18),
('0x220c1345fb076a6ef4dad9a9688e2b9b724528ff', 'UCHI', 9),
('0x24e79e946dea5482212c38aab2d0782f04cdb0e0', 'palStkAAVE', 18),
('0x3a9780555dac7ad1c1b252f466ddae09e5737f50', '⸙͎۪۫', 18),
('0x467719ad09025fcc6cf6f8311755809d45a5e5f3', 'AXL', 6),
('0x4fdbcc7331a03b1e323ccd8968966282950ad769', 'QMT', 18),
('0x6b0956258ff7bd7645aa35369b55b61b8e6d6140', 'LUCKY', 8),
('0x9625ce7753ace1fa1865a47aae2c5c2ce4418569', 'KAP', 18),
('0xb29edff8345b9d8039d8b9471c027d6e458a3c02', '$OOO', 18),
('0xb43b2a25330dda3366417a6a5f3b961154b6f99c', 'SUN', 18),
('0xc502002aeb1b9309fccb016adf50507987fc6c2b', 'GNFT', 18),
('0xc5bcc8ba3f33ab0d64f3473e861bdc0685b19ef5', '$MECHA', 18),
('0xe1e44e6918c283a0481037a03056b416f4347b28', 'DAZU', 9),
('0xe5f55a3b74874531a99359b833b92866a6609f6b', 'ATS', 4),
('0xe9f84d418b008888a992ff8c6d22389c2c3504e0', 'BASE', 8),
('0xfa372ff1547fa1a283b5112a4685f1358ce5574d', 'BDAMM', 18),
```

### CODE REVIEW

File: src/missing_tokens.py [hunk 1]
```diff
@@ -28,67 +27,75 @@ def __init__(self, address: Address):
...
         if version == DuneVersion.V2:
-            return f"('{self.address.lower()}', '{self.symbol}', {self.decimals}),"
+            return self.v2_string()
         raise ValueError(f"Invalid DuneVersion {version}")
 
+    def v1_string(self) -> str:
+        address_bytea = f"\\\\x{self.address[2:]}"
+        return f"{address_bytea}\t{self.symbol}\t{self.decimals}"
+
+    def v2_string(self) -> str:
```
Comment (gentrexha) [inline src/missing_tokens.py]: In my opinion, these method names were hard to get based only on their name. I think they could be a little bit more descriptive and follow the reference of Clean Code: 

>Methods should have verb or verb phrase names like postPayment , deletePage, or save.

ref: https://www.oreilly.com/library/view/clean-code-a/9780136083238/

File: src/missing_tokens.py [hunk 2]
```diff
@@ -28,67 +27,75 @@ def __init__(self, address: Address):
...
+
+@dataclass
+class MissingTokenResults:
+    v1: list[Address]
+    v2: list[Address]
+
+    def is_empty(self) -> bool:
+        return self.v1 is [] and self.v2 is []
+
+    def all_tokens(self) -> set[Address]:
```
Comment (gentrexha) [inline src/missing_tokens.py]: If we were to get nitpicky the same as above would ably here to: missing verb.

File: src/missing_tokens.py [hunk 3]
```diff
@@ -28,67 +27,75 @@ def __init__(self, address: Address):
...
-            return [Address(row["token"]) for row in csv.DictReader(csv_file)]
-
-    print("Getting missing tokens from: https://dune.com/queries/236085")
-    query = DuneQuery.from_environment(
-        raw_sql=open_query("./queries/missing-tokens.sql"),
-        name=f"Missing Tokens",
-        network=Network.MAINNET,
-        parameters=[],
+
+def fetch_missing_tokens_api(dune: DuneClient) -> MissingTokenResults:
```
Comment (gentrexha) [inline src/missing_tokens.py]: In my opinion if this going to be the default method to fetch missing tokens I think the `_api` unnecessary in this case.

File: src/missing_tokens.py [hunk 4]
```diff
@@ -28,67 +27,75 @@ def __init__(self, address: Address):
...
-                print(f"Something wrong with token {t} - skipping.")
-        print("\n".join(token_details))
+                print(f"Something wrong with token {token} - skipping.")
+
+        v1_results = "\n".join(token_details[t].v1_string() for t in missing_tokens.v1)
+        v2_results = "\n".join(token_details[t].v2_string() for t in missing_tokens.v2)
+
+        print(f"V1 results:\n\n{v1_results}\n")
+        print(f"V2 results:\n\n{v2_results}\n")
+        # TODO - write to file!
```
Comment (gentrexha) [inline src/missing_tokens.py:101]: There's still this TODO in here. Just mentioning this in case you wanted to do it and forgot.

**Review by gentrexha — APPROVED** [review]
Hey Ben, thanks for updating this! I've left some comments.

Questions:

- Is this is going to cost us 2$ every time we run it?

Requests:

- Could you please update the `.env.sample` file to include the `DUNE_API_KEY` env value?
- Could you also please update the `README.md` file for updated instructions on how to make use of the missing tokens script.

**Comment by bh2smith** [comment]
I think dune is reevaluating the api pricing plan and also, since this particular script is for improvement of dune data overall, I would argue that it should not be part of our billing plan.

File: src/missing_tokens.py (hunk 1 above)
Comment (bh2smith) [reply src/missing_tokens.py]: Sure. I'll try and think of something, but also open to suggestions.

File: src/missing_tokens.py (hunk 4 above)
Comment (bh2smith) [reply src/missing_tokens.py:101]: Actually I think I want to change the todo to be insert directly in spellbook abstraction file (which would be a separate task). For write to file I guess we could sneak it into this PR.

**Comment by bh2smith** [comment]
Fixed all the things. Up next - I will make this script fetch missing tokens for gnosis chain network (maybe eventually arbitrary networks).

File: src/missing_tokens.py (hunk 4 above)
Comment (bh2smith) [reply src/missing_tokens.py:101]: This later.

**Pull request merged**